Extensions/ArtifactEngine/Engine/artifactEngine.ts

```diff
@@ -8,37 +11,61 @@ import { ArtifactItemStore } from '../Store/artifactItemStore';
 import { ArtifactEngineOptions } from "./artifactEngineOptions"
 import { Logger } from './logger';
 import { Worker } from './worker';
+import { TicketState } from '../Models/ticketState';
+import { CacheProvider } from '../Providers/cacheProvider';
 
 export class ArtifactEngine {
     processItems(sourceProvider: models.IArtifactProvider, destProvider: models.IArtifactProvider, artifactEngineOptions?: ArtifactEngineOptions): Promise<models.ArtifactDownloadTicket[]> {
         var artifactDownloadTicketsPromise = new Promise<models.ArtifactDownloadTicket[]>((resolve, reject) => {
             const workers: Promise<void>[] = [];
             artifactEngineOptions = artifactEngineOptions || new ArtifactEngineOptions();
+            artifactEngineOptions.artifactCacheDirectory = artifactEngineOptions.artifactCacheDirectory ? artifactEngineOptions.artifactCacheDirectory : tl.getVariable("AGENT_WORKFOLDER");
             this.createPatternList(artifactEngineOptions);
+            var artifactName = sourceProvider.getRootItemPath();
+            this.artifactItemStore = new ArtifactItemStore();
             this.artifactItemStore.flush();
             Logger.verbose = artifactEngineOptions.verbose;
             this.logger = new Logger(this.artifactItemStore);
             this.logger.logProgress();
             sourceProvider.artifactItemStore = this.artifactItemStore;
             destProvider.artifactItemStore = this.artifactItemStore;
+            this.cacheProvider = new CacheProvider(artifactEngineOptions.artifactCacheDirectory,artifactEngineOptions.artifactCacheKey,artifactName);
             sourceProvider.getRootItems().then((itemsToProcess: models.ArtifactItem[]) => {
                 this.artifactItemStore.addItems(itemsToProcess);
-
-                for (let i = 0; i < artifactEngineOptions.parallelProcessingLimit; ++i) {
-                    var worker = new Worker<models.ArtifactItem>(i + 1, item => this.processArtifactItem(sourceProvider, item, destProvider, artifactEngineOptions), () => this.artifactItemStore.getNextItemToProcess(), () => !this.artifactItemStore.itemsPendingProcessing());
-                    workers.push(worker.init());
-                }
-
-                Promise.all(workers).then(() => {
-                    this.logger.logSummary();
-                    sourceProvider.dispose();
-                    destProvider.dispose();
-                    resolve(this.artifactItemStore.getTickets());
-                }, (err) => {
-                    ci.publishEvent('reliability', <ci.IReliabilityData>{ issueType: 'error', errorMessage: JSON.stringify(err, Object.getOwnPropertyNames(err)) });
-                    sourceProvider.dispose();
-                    destProvider.dispose();
-                    reject(err);
+                this.createNewHashMap(sourceProvider,itemsToProcess).then(() => {
```

Comment:
If createNewHashMap fails we should log a warning and continue with normal download